backend/sftp/sftp.go

```diff
@@ -226,6 +226,11 @@ E.g. the second example above should be rewritten as:
 			Default:  "",
 			Help:     "The command used to read sha1 hashes.\n\nLeave blank for autodetect.",
 			Advanced: true,
+		}, {
+			Name:     "blake3sum_command",
```

Comment:
Yes, I agree, and have noted your concern from an older comment on the TODO list :)

> I have a slight concern about sftp startup times. Lots of people use sftp without a config file which means that it probes for shells/supported hashes each time it is used. Perhaps we should delay hash support probing until it is asked for?
https://github.com/rclone/rclone/issues/7765#issuecomment-2057057464

I was briefly looking into the suggestion of delaying hash support probing, but found that it probably requires somewhat bigger refactoring in core parts that I did not want to dive into... If you mean adding a sftp-specific configuration parameter for enabling additional (or perhaps overriding default) hash types, more like the one I added for the local remote, then that would definitely be easier.

Reply:
> Also why blake3? It lost out to keccak in the SHA-3 contest.

I think because that was what was explicitly asked for in the issue, and also one I was familiar with, and know is in use (e.g. CCache uses both XXH3 and BLAKE3). Never heard of  Keccak, I must admit.

Reply:
> Yes, I agree, and have noted your concern from an older comment on the TODO list :)

Getting old and repeating myself ;-)

> > I have a slight concern about sftp startup times. Lots of people use sftp without a config file which means that it probes for shells/supported hashes each time it is used. Perhaps we should delay hash support probing until it is asked for?
> > [#7765 (comment)](https://github.com/rclone/rclone/issues/7765#issuecomment-2057057464)
> 
> I was briefly looking into the suggestion of delaying hash support probing, but found that it probably requires somewhat bigger refactoring in core parts that I did not want to dive into...

We already delay hash probing until `Fs.Hashes` is called which is good and doing anything else would require major surgery as you've noted.

Perhaps that is enough?

> If you mean adding a sftp-specific configuration parameter for enabling additional (or perhaps overriding default) hash types, more like the one I added for the local remote, then that would definitely be easier.

That might be a good idea.

Though we should probably inject some science into this - how long does the hash probing take?

```
$ rclone backend --log-format "microseconds" -vv features :sftp,host=localhost,user=sftptest: | jq .Hashes
11:21:56.689201 DEBUG : rclone: Version "v1.68.0-beta.8161.7194d828d.fix-4632-gofile" starting with parameters ["rclone" "backend" "--log-format" "microseconds" "-vv" "features" ":sftp,host=localhost,user=sftptest:"]
11:21:56.691707 DEBUG : Using config file from "/home/ncw/.rclone.conf"
11:21:56.740716 DEBUG : sftp://sftptest@localhost:22/: New connection 127.0.0.1:49344->127.0.0.1:22 to "SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.10"
11:21:57.083580 DEBUG : sftp://sftptest@localhost:22/: Running shell type detection remote command: echo ${ShellId}%ComSpec%
11:21:57.087933 DEBUG : sftp://sftptest@localhost:22/: Remote command result: %ComSpec%
11:21:57.087952 DEBUG : sftp://sftptest@localhost:22/: Shell type "unix" detected (set option shell_type to override)
11:21:57.087978 DEBUG : Saving config "shell_type" in section ":sftp" of the config file
11:21:57.088001 NOTICE: Can't save config "shell_type" for on the fly backend ":sftp"
11:21:57.097247 DEBUG : sftp://sftptest@localhost:22/: Relative path resolved to "/home/sftptest"
11:21:57.097263 DEBUG : sftp://sftptest@localhost:22/: Using root directory "/home/sftptest"
11:21:57.097302 DEBUG : sftp://sftptest@localhost:22/: Checking default md5 hash commands
11:21:57.097483 DEBUG : sftp://sftptest@localhost:22/: Running remote command: md5sum
11:21:57.102713 DEBUG : sftp://sftptest@localhost:22/: Remote command result: d41d8cd98f00b204e9800998ecf8427e  -
11:21:57.102740 DEBUG : sftp://sftptest@localhost:22/: Hash command accepted
11:21:57.102750 DEBUG : sftp://sftptest@localhost:22/: Checking default sha1 hash commands
11:21:57.102919 DEBUG : sftp://sftptest@localhost:22/: Running remote command: sha1sum
11:21:57.107854 DEBUG : sftp://sftptest@localhost:22/: Remote command result: da39a3ee5e6b4b0d3255bfef95601890afd80709  -
11:21:57.107885 DEBUG : sftp://sftptest@localhost:22/: Hash command accepted
11:21:57.107905 DEBUG : sftp://sftptest@localhost:22/: Setting hash command for md5 to "md5sum" (set sha1sum_command to override)
11:21:57.107921 DEBUG : Saving config "md5sum_command" in section ":sftp" of the config file
11:21:57.107945 NOTICE: Can't save config "md5sum_command" for on the fly backend ":sftp"
11:21:57.111972 DEBUG : sftp://sftptest@localhost:22/: Setting hash command for sha1 to "sha1sum" (set md5sum_command to override)
11:21:57.111991 DEBUG : Saving config "sha1sum_command" in section ":sftp" of the config file
11:21:57.112012 NOTICE: Can't save config "sha1sum_command" for on the fly backend ":sftp"
11:21:57.117542 DEBUG : 13 go routines active
[
  "md5",
  "sha1"
]
```

vs

```
$ rclone backend --log-format "microseconds" -vv features :sftp,host=localhost,user=sftptest,disable_hashcheck: | jq .Hashes
11:23:14.557982 DEBUG : rclone: Version "v1.68.0-beta.8161.7194d828d.fix-4632-gofile" starting with parameters ["rclone" "backend" "--log-format" "microseconds" "-vv" "features" ":sftp,host=localhost,user=sftptest,disable_hashcheck:"]
11:23:14.560598 DEBUG : Using config file from "/home/ncw/.rclone.conf"
11:23:14.607004 DEBUG : sftp://sftptest@localhost:22/: New connection 127.0.0.1:52498->127.0.0.1:22 to "SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.10"
11:23:14.963461 DEBUG : sftp://sftptest@localhost:22/: Running shell type detection remote command: echo ${ShellId}%ComSpec%
11:23:14.967806 DEBUG : sftp://sftptest@localhost:22/: Remote command result: %ComSpec%
11:23:14.967821 DEBUG : sftp://sftptest@localhost:22/: Shell type "unix" detected (set option shell_type to override)
11:23:14.967835 DEBUG : Saving config "shell_type" in section ":sftp" of the config file
11:23:14.967857 NOTICE: Can't save config "shell_type" for on the fly backend ":sftp"
11:23:14.977497 DEBUG : sftp://sftptest@localhost:22/: Relative path resolved to "/home/sftptest"
11:23:14.977514 DEBUG : sftp://sftptest@localhost:22/: Using root directory "/home/sftptest"
11:23:14.977763 DEBUG : 13 go routines active
[]
```

So checking 2 hashes went from 419 ms without to 425 ms with on a local machine so 3 ms per hash

The same test on a remote machine gives 703 ms vs 957 ms which is probably more typical given the round trip times. So roughly 120 ms per hash checked. This server is 40 ms ping away so indicating that is 3 round trips.

Could we detect this quicker? Maybe we could do something like

```
$ which md5sum md5 rclone sha1sum sha
/usr/bin/md5sum
/home/ncw/go/bin//rclone
/usr/bin/sha1sum
```

To detect whether all the binaries exist at once? And if the binary exists don't bother to test it. Or only test the ones which actually existed.

Or we could send all the test commands at once, effectively sending a shell script something like

```
$ ssh sftptest@localhost "(md5sum ; md5 ; sha1sum ; sha1 ; rclone md5sum ; rclone sha1sum )</dev/null"
d41d8cd98f00b204e9800998ecf8427e  -
bash: line 1: md5: command not found
da39a3ee5e6b4b0d3255bfef95601890afd80709  -
bash: line 1: sha1: command not found
bash: line 1: rclone: command not found
bash: line 1: rclone: command not found
```

(I think the `</dev/null` is implicit in the way we run these commands).